### PR TITLE
feat(utils): add an in-memory and a fs-backed cache

### PR DESCRIPTION
**Pull request by sidvishnoi**
In hopes of having a consistent cache interface across modules.. 
Right now, each route has its own cache.

### CODE REVIEW

File: utils/fs-cache.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+      return fromMemory;
+    }
+    this.#deleteQueue.push(key);
+
+    try {
+      /** @type {FsCacheEntry} */
+      const { time, value } = await readJSON(this.getFilePath(key));
+      this.#cache._set(key, time, value);
+      return value;
+    } catch {
```
Comment (marcoscaceres) [inline utils/fs-cache.js:58]: you sure you want this silent?

File: utils/fs-cache.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+  /**
+   * @param {string} dirname where to store cached items on file-system
+   * @typedef {{ key: string, time: number, value: ValueType }} FsCacheEntry
+   */
+  constructor(dirname, ttl = Infinity, autosaveInterval = 10000) {
+    this.#dirname = dirname;
+    this.#cache = new MemCache(ttl);
+    if (autosaveInterval > 0) {
+      setTimeout(this.dump.bind(this), autosaveInterval);
```
Comment (marcoscaceres) [inline utils/fs-cache.js:30]: Just wondering, what's the rationale of using the deleted file count here?  seems a bit arbitrary is all.

File: utils/fs-cache.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+   * Clears stale entries from in-memory cache as well as disk.
+   */
+  async invalidate() {
+    const invalidatedKeys = this.#cache.invalidate();
+    const keysToDelete = [
+      ...new Set([...invalidatedKeys, ...this.#deleteQueue]),
+    ];
+    this.#deleteQueue.length = 0;
+    const filesToDelete = keysToDelete.map(key => this.getFilePath(key));
+    await Promise.all(filesToDelete.map(file => unlink(file).catch(() => {})));
```
Comment (marcoscaceres) [inline utils/fs-cache.js:100]: same as above... seems like hiding errors might not be good.

File: utils/mem-cache.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+  constructor(ttl) {
+    this.#ttl = ttl;
+  }
+
+  /**
+   * @param {string} key
+   * @param {ValueType} value
+   */
+  set(key, value) {
+    this._set(key, Date.now(), value);
```
Comment (marcoscaceres) [inline utils/mem-cache.js:23]: the private `._set` and `._get` seem to just add unnecessary indirection (and yucky python looking... omg so gross!). Maybe consider inlining them? 🙏

File: utils/fs-cache.js (hunk 1 above)
Comment (sidvishnoi) [reply utils/fs-cache.js:58]: I think yes. It means, the key was neither in memory nor in file-system. file-system is more like "it's good if it's there, else we'll wait for a `set` anyway"

File: utils/fs-cache.js (hunk 2 above)
Comment (sidvishnoi) [reply utils/fs-cache.js:30]: Um.. this comment seems to be out of context.

File: utils/mem-cache.js (hunk 4 above)
Comment (sidvishnoi) [reply utils/mem-cache.js:23]: Used them here we're using them in fs-cache also..

**Comment by sidvishnoi** [comment]
On a second thought, I think I'll skip fs-cache in this PR, until I have a concrete use case for it in rest of the code. WDYT? Thanks for reviewing though, it'll be helpful when we add it back again

File: utils/fs-cache.js (hunk 2 above)
Comment (marcoscaceres) [reply utils/fs-cache.js:30]: oops, sorry... I read the function backwards... the correct question is, why does .dump() return a number?

**Comment by marcoscaceres** [comment]
> On a second thought, I think I'll skip fs-cache in this PR, until I have a concrete use case for it in rest of the code. WDYT?

sounds good to me.

**Pull request merged**